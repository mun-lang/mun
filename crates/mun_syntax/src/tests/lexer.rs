use std::fmt::Write;

fn dump_tokens(tokens: &[crate::Token], text: &str) -> String {
    let mut acc = String::new();
    let mut offset = 0;
    for token in tokens {
        let len: u32 = token.len.into();
        let len = len as usize;
        let token_text = &text[offset..offset + len];
        offset += len;
        write!(acc, "{:?} {} {:?}\n", token.kind, token.len, token_text).unwrap()
    }
    acc
}

fn lex_snapshot(text: &str) {
    let text = text.trim().replace("\n    ", "\n");
    let tokens = crate::tokenize(&text);
    insta::assert_snapshot!(
        insta::_macro_support::AutoName,
        dump_tokens(&tokens, &text),
        &text
    );
}

#[test]
fn numbers() {
    lex_snapshot(
        r#"
    1.34
    0x3Af
    1e-3
    100_000"#,
    )
}

#[test]
fn comments() {
    lex_snapshot(
        r#"
    // hello, world!
    /**/
    /* block comment */
    /* multi
       line
       comment */
    /* /* nested */ */
    /* unclosed comment"#,
    )
}

#[test]
fn whitespace() {
    lex_snapshot(
        r#"
    h e ll  o
    w

    o r     l   d"#,
    )
}

#[test]
fn ident() {
    lex_snapshot(
        r#"
    hello world_ _a2 _ __ x 即可编著课程
    "#,
    )
}

#[test]
fn symbols() {
    lex_snapshot(
        r#"
    # ( ) { } [ ] ; ,
    = ==
    !=
    < <=
    > >=
    . .. ... ..=
    + +=
    - -=
    * *=
    / /=
    ^ ^=
    % %=
    : ::
    "#,
    )
}

#[test]
fn strings() {
    lex_snapshot(
        r#"
    "Hello, world!"
    'Hello, world!'
    "\n"
    "\"\\"
    "multi
    line"
    "#,
    )
}

#[test]
fn keywords() {
    lex_snapshot(
        r#"
    and break do else false for fn if in nil
    return true while let mut class public protected
    private
    "#,
    )
}

#[test]
fn unclosed_string() {
    lex_snapshot(
        r#"
    "test
    "#,
    )
}

#[test]
fn binary_cmp() {
    lex_snapshot(
        r#"
    a==b
    a ==b
    a== b
    a == b
    "#,
    )
}
