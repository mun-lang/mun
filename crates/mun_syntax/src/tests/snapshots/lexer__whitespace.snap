---
source: crates/mun_syntax/src/tests/lexer.rs
expression: "h e ll  o\nw\n\no r     l   d"
---
IDENT 1 "h"
WHITESPACE 1 " "
IDENT 1 "e"
WHITESPACE 1 " "
IDENT 2 "ll"
WHITESPACE 2 "  "
IDENT 1 "o"
WHITESPACE 1 "\n"
IDENT 1 "w"
WHITESPACE 2 "\n\n"
IDENT 1 "o"
WHITESPACE 1 " "
IDENT 1 "r"
WHITESPACE 5 "     "
IDENT 1 "l"
WHITESPACE 3 "   "
IDENT 1 "d"

